crypto/ledger_integration_test.go

```diff
@@ -0,0 +1,163 @@
+
+package crypto
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/cosmos/cosmos-sdk/crypto/keys/hd"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/stretchr/testify/require"
+	"github.com/tendermint/tendermint/crypto/encoding/amino"
+	ledger "github.com/zondax/ledger-cosmos-go"
+)
+
+const (
+	// These tests expect a ledger device initialized to the following mnemonic
+	testMnemonic = "equip will roof matter pink blind book anxiety banner elbow sun young"
+)
+
+func TestDiscoverDevice(t *testing.T) {
+	device, err := discoverLedger()
+	require.NoError(t, err)
+	require.NotNil(t, device)
+	defer device.Close()
+}
+
+func TestDiscoverDeviceTwice(t *testing.T) {
+	// We expect the second call not to find a device
+	device, err := discoverLedger()
+	require.NoError(t, err)
+	require.NotNil(t, device)
+	defer device.Close()
+
+	device2, err := discoverLedger()
+	require.Error(t, err)
+	require.Equal(t, "no ledger connected", err.Error())
+	require.Nil(t, device2)
+}
+
+func TestDiscoverDeviceTwiceClosing(t *testing.T) {
+	{
+		device, err := ledger.FindLedgerCosmosUserApp()
+		require.NoError(t, err)
+		require.NotNil(t, device)
+		require.NoError(t, device.Close())
+	}
+
+	device2, err := discoverLedger()
+	require.NoError(t, err)
+	require.NotNil(t, device2)
+	require.NoError(t, device2.Close())
+}
+
+func TestPublicKey(t *testing.T) {
+	path := *hd.NewFundraiserParams(0, 0)
+	priv, err := NewPrivKeyLedgerSecp256k1(path)
+	require.Nil(t, err, "%s", err)
+	require.NotNil(t, priv)
+
+	pubKeyAddr, err := sdk.Bech32ifyAccPub(priv.PubKey())
+	require.NoError(t, err)
+	require.Equal(t, "cosmospub1addwnpepqd87l8xhcnrrtzxnkql7k55ph8fr9jarf4hn6udwukfprlalu8lgw0urza0", pubKeyAddr, "Is your device using test mnemonic: %s ?", testMnemonic)
+	require.Equal(t, "5075624b6579536563703235366b317b303334464546394344374334433633353838443342303"+
+		"3464542353238314239443233324342413334443646334437314145453539323131464642464531464538377d",
+		fmt.Sprintf("%x", priv.PubKey()))
+}
+
+func TestPublicKeyHDPath(t *testing.T) {
+	expectedAnswers := []string{
+		"cosmospub1addwnpepqd87l8xhcnrrtzxnkql7k55ph8fr9jarf4hn6udwukfprlalu8lgw0urza0",
+		"cosmospub1addwnpepqfsdqjr68h7wjg5wacksmqaypasnra232fkgu5sxdlnlu8j22ztxvlqvd65",
+		"cosmospub1addwnpepqw3xwqun6q43vtgw6p4qspq7srvxhcmvq4jrx5j5ma6xy3r7k6dtxmrkh3d",
+		"cosmospub1addwnpepqvez9lrp09g8w7gkv42y4yr5p6826cu28ydrhrujv862yf4njmqyyjr4pjs",
+		"cosmospub1addwnpepq06hw3enfrtmq8n67teytcmtnrgcr0yntmyt25kdukfjkerdc7lqg32rcz7",
+		"cosmospub1addwnpepqg3trf2gd0s2940nckrxherwqhgmm6xd5h4pcnrh4x7y35h6yafmcpk5qns",
+		"cosmospub1addwnpepqdm6rjpx6wsref8wjn7ym6ntejet430j4szpngfgc20caz83lu545vuv8hp",
+		"cosmospub1addwnpepqvdhtjzy2wf44dm03jxsketxc07vzqwvt3vawqqtljgsr9s7jvydjmt66ew",
+		"cosmospub1addwnpepqwystfpyxwcava7v3t7ndps5xzu6s553wxcxzmmnxevlzvwrlqpzz695nw9",
+		"cosmospub1addwnpepqw970u6gjqkccg9u3rfj99857wupj2z9fqfzy2w7e5dd7xn7kzzgkgqch0r",
+	}
+
+	// TODO: Check data with locally generated pubkey
```

Comment:
wrt to this commented block, can we we remove it? Is this PR still dependent on #3461?

Reply:
The test is fine as it is but code in #3461 allows for a bit more extensive testing.
I will remove those comments. 

We still need more testing anyway as described in https://github.com/cosmos/cosmos-sdk/issues/2004 This can be done there.